api/repositories/app_repository.go

```diff
@@ -156,22 +161,70 @@ func (f *AppRepo) CreateApp(ctx context.Context, client client.Client, appCreate
 	return cfAppToAppRecord(cfApp), err
 }
 
-func (f *AppRepo) FetchAppList(ctx context.Context, client client.Client) ([]AppRecord, error) {
+func (f *AppRepo) FetchAppList(ctx context.Context, client client.Client, message AppListMessage) ([]AppRecord, error) {
 	// TODO: add checks for allowed namespaces with completion of initial auth work
 	// Currently we assume the client has full cluster access and naively returns all records
 	appList := &workloadsv1alpha1.CFAppList{}
 	err := client.List(ctx, appList)
 	if err != nil {
 		return []AppRecord{}, err
 	}
-	allApps := appList.Items
+	filteredApps := f.applyAppListFilter(appList.Items, message)
+
+	return f.returnAppList(filteredApps), nil
+}
+
+func (f *AppRepo) applyAppListFilter(appList []workloadsv1alpha1.CFApp, message AppListMessage) []workloadsv1alpha1.CFApp {
+	nameFilterSpecified := len(message.Names) > 0
+	spaceGUIDFilterSpecified := len(message.SpaceGuids) > 0
 
-	appRecordList := make([]AppRecord, 0, len(allApps))
-	for _, app := range allApps {
-		appRecordList = append(appRecordList, cfAppToAppRecord(app))
+	if !nameFilterSpecified && !spaceGUIDFilterSpecified {
+		return appList
+	}
+
+	var filtered []workloadsv1alpha1.CFApp
+	for _, app := range appList {
+		if nameFilterSpecified && spaceGUIDFilterSpecified {
+			for _, name := range message.Names {
+				for _, spaceGUID := range message.SpaceGuids {
+					if appBelongsToSpace(app, spaceGUID) && appMatchesName(app, name) {
+						filtered = append(filtered, app)
+					}
+				}
+			}
+		} else if nameFilterSpecified {
+			for _, name := range message.Names {
+				if appMatchesName(app, name) {
+					filtered = append(filtered, app)
+				}
+			}
+		} else if spaceGUIDFilterSpecified {
+			for _, spaceGUID := range message.SpaceGuids {
+				if appBelongsToSpace(app, spaceGUID) {
+					filtered = append(filtered, app)
+				}
+			}
+		}
 	}
```

Comment:
I'm not too happy with the logic I came up with here to try and perform the filter union.

The most natural thing IMO would be to start with the whole list and remove that don't match each particular filter. Since Go doesn't have a built-in remove I implemented my own and saw how many times I was iterating over the collection of apps and just settled with this for now.

As we add more filtering and more complex filters across endpoints we might want to revisit this. Maybe use generics 😈

Reply:
This seems fine for now. 
For the CRD explore we took a similar approach as you described:
https://github.com/cloudfoundry/cf-crd-explorations/blob/8cf10d5aafec5695394c1f4eebf21ce481ca800b/cfshim/filters/app_filter.go#L13